mesh/mesh.go

```diff
@@ -436,7 +463,7 @@ func (msh *Mesh) reInsertTxsToPool(validBlocks, invalidBlocks []*types.Block, l
 	uniqueTxIds(validBlocks, seenTxIds) // run for the side effect, updating seenTxIds
 	returnedTxs, missing := msh.GetTransactions(uniqueTxIds(invalidBlocks, seenTxIds))
 	if len(missing) > 0 {
-		msh.Panic("could not find transactions %v from layer %v", missing, l)
```

Comment:
but those are from the invalid blocks, how do you know that the invalid block actually referenced existing transactions and not some random ids?

Reply:
I might be wrong about this, but I thought that blocks go through syntactic validation when they're received. Syntactic validation should include data availability.

This specific code is not where we handle an incoming block (be it via gossip or sync). It's not where we handle a syntactically invalid block (that might include garbage) - rather it's where we handle a syntactically valid block that's been rejected by the consensus mechanism. At this point we should have already obtained the dependencies. If we couldn't obtain them (because, as you said someone may have created a block with garbage, non-existent transactions) then it wouldn't be syntactically valid and any object referencing it (e.g. a ballot) shouldn't be considered syntactically valid, either.

If I got it wrong then we should discuss this to make sure that the behavior makes sense. We eventually want to support pruning transactions, but this is not as simple as just not having them when they're referenced in a block (even a contextually invalid one - those are precisely the transactions we would want to prune).

Reply:
in general yes. but right now there is no strict check for transactions availability (i don't know for how long this was the case). fetcher just tries to download transactions, if it can't it moves on. and i don't think that we should change that.

in my understanding that in the long term when we receive a block we won't be able to determine if it was pruned or simply malicious being withheld by a peer. so we can't simply ignore it. we also can't try to download it indefinitely.  instead node should count votes for that block and if according to the tortoise such block is valid - we can assume that the block is available and can be downloaded.

this is why i think that we should relax data availability for blocks during gossip and enforce it only when we know which blocks are valid.

Reply:
This code deals with proposals, not blocks (despite the outdated naming) and a proposal is only ever stored by gossiping:

The only production code path that leads to storing a proposal goes through `HandleProposalData` and there's this check there that prevents storing the proposal if a transaction is unavailable:
https://github.com/spacemeshos/go-spacemesh/blob/c75fc000224636e881f48999245d7dc3a5447615/proposals/handler.go#L207-L210

Reply:
you should look into fetcher.GetTXs